src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,130 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int arraySize;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private T[] grow() {
+        return (T[]) new Object[elementData.length + elementData.length / 2];
+    }
 
     @Override
     public void add(T value) {
-
+        if (elementData.length > arraySize + 1) {
+            elementData[arraySize++] = value;
+        } else {
+            T[] newElementData = grow();
+            System.arraycopy(elementData, 0, newElementData,
+                            0, arraySize);
+            newElementData[arraySize++] = value;
+            elementData = newElementData;
+        }
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > arraySize) {
+            throw new ArrayListIndexOutOfBoundsException("Index is invalid");
+        }
```

Comment:
this logic is the same in many methods, let's move it to the separate private method, which will get 2 parameters: int index and int maxSize

Reply:
done